h/views/api/group_members.py

```diff
@@ -54,3 +60,33 @@ def add_member(context: GroupMembershipContext, request):
     group_members_service.member_join(context.group, context.user.userid)
 
     return HTTPNoContent()
+
+
+@api_config(
+    versions=["v1", "v2"],
+    route_name="api.group_member",
+    request_method="PATCH",
+    link_name="group.member.edit",
+    description="Change a user's role in a group",
+)
+def edit_member(context: GroupMembershipContext, request):
+    appstruct = EditGroupMembershipAPISchema().validate(json_payload(request))
+    old_roles = context.membership.roles
+    new_roles = appstruct["roles"]
+
+    if not request.has_permission(
+        Permission.Group.MEMBER_EDIT,
+        EditGroupMembershipContext(
+            context.group, context.user, context.membership, new_roles
+        ),
+    ):
+        raise HTTPNotFound()
```

Comment:
There's no `permission` in the `@api_config()` on this view. The reason is that whether or not you have permission to change a role depends on the new role that you want to change it to (only owners can change someone's role to owner or admin, admins can change someone's role to moderator or member, etc).

The new role comes from the request JSON body and isn't available until that JSON body has been parsed and validated, and that happens in the view. So we can't use Pyramid's `permission` thing where it calls the security policy before calling the view.

Instead we let Pyramid call the view regardless, and then the view itself constructs a context object and calls `request.has_permission()`.

We do need the permissions logic to be implemented in the centralized permissions system because in future another view is going to have to check `MEMBER_EDIT` permissions as well, so we couldn't just code the logic directly in this view. Also permissions logic just belongs in the permissions system.